File: 279-Perfect_Squares/main.go
package main

import (
	"fmt"
	"math"
)

// 自顶向下的方案
func numSquares(n int) int {
	res := make([]int, n+1)

	for i := 1; i <= n; i++ {
		res[i] = -1
	}

	return dp(n, res)
}

func dp(n int, res []int) int {
	if res[n] != -1 {
		return res[n]
	}

	min := math.MaxInt32

	for i := 1; i*i <= n; i++ {
		if 1 + dp(n - i*i, res) < min {
			min = 1 + dp(n - i*i, res)
		}
	}

	res[n] = min
	return min
}

func main() {
	fmt.Println(numSquares(13))
}

File: 279-Perfect_Squares/main.js
/**
 * 给定正整数 n，找到若干个完全平方数（比如 1, 4, 9, 16, ...）使得它们的和等于 n。
 * 你需要让组成和的完全平方数的个数最少。
 * 自底向上的动态规划方案
 * 这里的陷阱是在 s[0] 处，出现了 s[0] 说明当前这个数就是一个平方数
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
  let s = Array(n+1);
  s[0] = 0;
  s[1] = 1;

  for (let i = 2; i <= n; i++) {
    let min = +Infinity;
    for (let j = 1; j*j <= i; j++) {
      if (1 + s[i - j*j] < min) {
        min = 1 + s[i- j*j];
      }
    }
    s[i] = min;
  }

  return s[n];
};

console.log(numSquares(12));